controller/src/main/java/io/pravega/controller/store/kvtable/InMemoryKVTMetadataStore.java

```diff
@@ -36,48 +36,44 @@
     @GuardedBy("$lock")
     private final Map<String, Integer> deletedKVTables = new HashMap<>();
 
-    @Setter
-    @GuardedBy("$lock")
-    private Map<String, InMemoryScope> scopes = new HashMap<>();
-
-    private final AtomicInt96 counter;
-
-    private final Executor executor;
+    private final InMemoryStreamMetadataStore streamStore;
+    private final ScheduledExecutorService executor;
 
-    public InMemoryKVTMetadataStore(Executor executor) {
+    public InMemoryKVTMetadataStore(StreamMetadataStore streamStore, ScheduledExecutorService executor) {
         super(new InMemoryHostIndex());
+        this.streamStore = (InMemoryStreamMetadataStore) streamStore;
         this.executor = executor;
-        this.counter = new AtomicInt96();
     }
 
     @Override
-    KeyValueTable newKeyValueTable(String scope, String kvTableName) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public KeyValueTable getKVTable(String scope, final String name, KVTOperationContext context) {
-        KeyValueTable kvt;
-        if (context != null) {
-            kvt = context.getKvTable();
-            assert kvt.getScopeName().equals(scope);
-            assert kvt.getName().equals(name);
-        } else {
-            if (!scopes.containsKey(scope)) {
-                return new InMemoryKVTable(scope, name);
+    @Synchronized
+    KeyValueTable newKeyValueTable(String scope, String name) {
+        if (this.streamStore.getScopes().containsKey(scope)) {
+            InMemoryScope kvtScope = this.streamStore.getScopes().get(scope);
```

Comment:
streamstore.getScopes is returning a map which is not synchronization protected in its access here.. 
instead expose a syncrhonized method on the stream store for getting the scope.

Reply:
Sure.

Reply:
fixed

Reply:
creating an immutable map each time its accessed here is costly as it will make a copy. 
instead you can use `Collections.unmodifiableMap` which actually creates a wrapper decoration on the map which prohibits any changes to the map

Reply:
actually i dont see you returning an immutable map.

immutable map is one in which you cannot add/remove/replace any entries. 

setting the field as final doesnt make it immutable, it just means that field cannot be assigned a different value. but the object that the field refers to can still be immutable, as is the case here.

Reply:
Changed as discussed.